stacker/plan.py

```diff
@@ -97,174 +86,79 @@ def submit(self):
-    steps in the order they are added to the `Plan` via the :func:`Plan.add`
-    function. If a `Step` specifies requirements, the `Plan` will wait until
-    the required stacks have completed before executing that `Step`.
-
-    Args:
-        description (str): description of the plan
-        sleep_time (int, optional): the amount of time that will be passed to
-            the `wait_func`. Default: 5 seconds.
-        wait_func (func, optional): the function to be called after each pass
-            of running stacks. This defaults to :func:`time.sleep` and will
-            sleep for the given `sleep_time` before starting the next pass.
-            Default: :func:`time.sleep`
-
-    """
-
-    def __init__(self, description, sleep_time=5, wait_func=None,
-                 watch_func=None, *args, **kwargs):
+class Plan():
+    def __init__(self, description, reverse=False, sleep_func=sleep):
         self.description = description
-        self.sleep_time = sleep_time
-        if wait_func is not None:
-            if not callable(wait_func):
-                raise ImproperlyConfigured(self.__class__,
-                                           "\"wait_func\" must be a callable")
-            self._wait_func = wait_func
-        else:
-            self._wait_func = time.sleep
-
-        self._watchers = {}
-        self._watch_func = watch_func
+        self._dag = None
+        self._steps = {}
+        self._reverse = reverse
+        self._sleep_func = sleep_func
         self.id = uuid.uuid4()
-        super(Plan, self).__init__(*args, **kwargs)
-
-    def add(self, stack, run_func, requires=None):
-        """Add a new step to the plan.
-
-        Args:
-            stack (:class:`stacker.stack.Stack`): The stack to add to the plan.
-            run_func (function): The function to call when the step is ran.
-            requires (list, optional): A list of other stacks that are required
-                to be complete before this step is started.
-        """
-        self[stack.fqn] = Step(
-            stack=stack,
-            run_func=run_func,
-            requires=requires,
-        )
 
-    def list_status(self, status):
-        """Returns a list of steps in the given status.
-
-        Args:
-            status (:class:`Status`): The status to match steps against.
-
-        Returns:
-            list: A list of :class:`Step` objects that are in the given status.
+    def build(self, stacks):
+        """ Builds an internal dag from the stacks and their dependencies """
+        dag = DAG()
+
+        for stack in stacks:
+            fqn = stack.fqn
+            dag.add_node(fqn)
+            self._steps[fqn] = Step(
+                stack=stack)
+
+        for stack in stacks:
+            for dep in stack.requires:
+                try:
+                    dag.add_edge(stack.fqn, dep)
+                except DAGValidationError:
+                    raise CyclicDependencyError(stack.fqn)
```

Comment:
You should add the dependency to this exception as well, just so it's easier to see what dependency caused the error.